### PR TITLE
feat: create GET /providers/:id to response user profile information

### PR DESCRIPTION
**Pull request by devstm**
Relates #29
end point: /providers/:id

create controller function to get all user information
- user's personal information
- works
- locations
- reviews
- services

create validation schema to validate req.params:
- id must be number, integer, positive and require.

create 10 passed tests to this end point to check is returned data correct  with status 200 
or data incorrect with status 400
not found user with status 404 and 500 

data responded:
localhost_3030_api_v1_providers_2
9

### CODE REVIEW

**Review by abdulrahman-2020 — APPROVED** [review]

**Review by RandSohail — APPROVED** [review]

**Review by AbdallahGot — APPROVED** [review]

File: server/controllers/user/profile.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    const { id } = await paramsValidation.validateAsync(req.params);
+    const works = await Work.findAll({
+      where: {
+        userId: id,
+      },
+    });
+    const user = await works[0].getUser();
+    const services = await user.getServices();
+    const locations = await user.getLocation();
+    const reviews = await Review.findAll({
```
Comment (LinaYahya) [inline server/controllers/user/profile.js:16]: Do we need to get all review we only need to get the top five or ten reviews,
The same for works I think in the first fetch we need to get the most recent four or six works as we have in figma, then when user navigate to the next page we can fetch more

And as we're not getting the whole reviewsm we still need the whole rate for that provider

File: server/controllers/user/profile.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  try {
+    const { id } = await paramsValidation.validateAsync(req.params);
+    const works = await Work.findAll({
+      where: {
+        userId: id,
+      },
+    });
+    const user = await works[0].getUser();
+    const services = await user.getServices();
+    const locations = await user.getLocation();
```
Comment (LinaYahya) [inline server/controllers/user/profile.js:15]: can you imagine how many await we have here,
Why not trying to fetch data using include to join tables we have
```js
await User.findByPk(id, {
      include: [{ model: Location }, { model: SubServices }],
      })
 ```
 So what if user here will be null ?! Do we need to send status 200, I doubt that
 So we need to check if user exists then maybe we can get reviews and works after that
 and while neither review nor work depends on each other we can run queries at the same time using `Promise.all`
 
 And did you notice that we're sending passwords to the client-side!! 😨 
 Also, it's good to not sending all raws even we don't use, So try to exclude columns we don't need

File: server/test/testData.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
+const testData = {
```
Comment (LinaYahya) [inline server/test/testData.js:1]: we don't need to test routes on fake data,
we need to build fake data first as we did before
so afterAll tests run build 
and after all test close connection

File: server/utils/validation/params.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+// this file to validate params.id attribute
```
Comment (LinaYahya) [inline server/utils/validation/params.js:1]: rename file name 
validation/params not explain anythings

**Review by LinaYahya — CHANGES_REQUESTED** [review]

File: server/controllers/user/profile.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      include: [{ model: Location }, { model: SubServices }],
+      attributes: {
+        exclude: ['password', 'createdAt', 'updatedAt'],
+      },
+    });
+    if (!user) {
+      throw customError('user does not exist', 400);
+    }
+    const reviewsAndWorks = await Promise.all([
+      await Review.findAll({
```
Comment (LinaYahya) [inline server/controllers/user/profile.js:26]:
```suggestion
       Review.findAll({
```

File: server/routes/index.js [hunk 6]
```diff
@@ -1,4 +1,7 @@
 const router = require('express').Router();
+const { profile } = require('../controllers/user');
+
+router.get('/providers/:id', profile);
```
Comment (LinaYahya) [inline server/routes/index.js:4]:
```suggestion
router.get('/providers/:id', getProviderInfo);
```

File: server/test/profile.test.js [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+      .expect(404)
+      .end((err, res) => {
+        if (err) done(err);
+        expect(res.body.status).toEqual(404);
+        done();
+      });
+  });
+  test('should first', () => {
+    expect(1).toBe(1);
+  });
```
Comment (LinaYahya) [inline server/test/profile.test.js:96]: remove that test and add a test to test not found user

**Review by LinaYahya — COMMENTED** [review]
Good job Saleh, just add a total rate for that provider
and works page count

File: server/controllers/user/profile.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        },
+      }),
+      await Work.findAll({
+        limit: 10,
+        where: {
+          userId: id,
+        },
+      }),
+    ]);
+    console.log(fn);
```
Comment (LinaYahya) [inline server/controllers/user/profile.js:42]: remove that line

File: server/controllers/user/profile.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    const totalReviews = await Review.sum('rate', { where: { userId: id } }); // 50
+    const count = await Review.count({ where: { userId: id } }); // 50
```
Comment (LinaYahya) [inline server/controllers/user/profile.js:44]: use promise all

**Review by LinaYahya — CHANGES_REQUESTED** [review]

**Pull request merged**